File: services/ai-service/src/models/conversation.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{FromRow, types::Json};
use uuid::Uuid;

use crate::models::knowledge_base::KnowledgeSearchResult;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GuardrailFlag {
    pub kind: String,
    pub severity: String,
    pub excerpt: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GuardrailVerdict {
    pub status: String,
    pub redacted_text: String,
    pub blocked: bool,
    pub flags: Vec<GuardrailFlag>,
}

impl Default for GuardrailVerdict {
    fn default() -> Self {
        Self {
            status: "passed".to_string(),
            redacted_text: String::new(),
            blocked: false,
            flags: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatAttachment {
    #[serde(default = "default_attachment_kind")]
    pub kind: String,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub mime_type: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub base64_data: Option<String>,
    #[serde(default)]
    pub text: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
    pub provider_id: Option<Uuid>,
    pub tool_name: Option<String>,
    #[serde(default)]
    pub citations: Vec<KnowledgeSearchResult>,
    #[serde(default)]
    pub attachments: Vec<ChatAttachment>,
    pub guardrail_verdict: Option<GuardrailVerdict>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LlmUsageSummary {
    pub prompt_tokens: i32,
    pub completion_tokens: i32,
    pub total_tokens: i32,
    pub estimated_cost_usd: f32,
    pub latency_ms: i32,
    pub network_scope: String,
    pub cache_hit: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatRoutingMetadata {
    pub requested_private_network: bool,
    pub used_private_network: bool,
    pub privacy_reason: Option<String>,
    pub candidate_provider_ids: Vec<Uuid>,
    pub required_modalities: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SemanticCacheMetadata {
    pub cache_key: String,
    pub hit: bool,
    pub similarity_score: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Conversation {
    pub id: Uuid,
    pub title: String,
    pub messages: Vec<ChatMessage>,
    pub provider_id: Option<Uuid>,
    pub last_cache_hit: bool,
    pub last_guardrail_blocked: bool,
    pub created_at: DateTime<Utc>,
    pub last_activity_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationSummary {
    pub id: Uuid,
    pub title: String,
    pub last_message_preview: String,
    pub provider_id: Option<Uuid>,
    pub message_count: i32,
    pub last_cache_hit: bool,
    pub last_activity_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListConversationsResponse {
    pub data: Vec<ConversationSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionRequest {
    pub conversation_id: Option<Uuid>,
    pub system_prompt: Option<String>,
    pub user_message: String,
    pub prompt_template_id: Option<Uuid>,
    #[serde(default)]
    pub prompt_variables: Value,
    pub knowledge_base_id: Option<Uuid>,
    pub preferred_provider_id: Option<Uuid>,
    #[serde(default)]
    pub attachments: Vec<ChatAttachment>,
    #[serde(default = "default_fallback_enabled")]
    pub fallback_enabled: bool,
    #[serde(default)]
    pub require_private_network: bool,
    #[serde(default = "default_temperature")]
    pub temperature: f32,
    #[serde(default = "default_max_tokens")]
    pub max_tokens: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionResponse {
    pub conversation_id: Uuid,
    pub provider_id: Uuid,
    pub provider_name: String,
    pub reply: String,
    pub citations: Vec<KnowledgeSearchResult>,
    pub guardrail: GuardrailVerdict,
    pub cache: SemanticCacheMetadata,
    pub prompt_used: String,
    pub completion_tokens: i32,
    pub usage: LlmUsageSummary,
    pub routing: ChatRoutingMetadata,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CopilotRequest {
    pub question: String,
    #[serde(default)]
    pub dataset_ids: Vec<Uuid>,
    #[serde(default)]
    pub ontology_type_ids: Vec<Uuid>,
    #[serde(default)]
    pub knowledge_base_ids: Vec<Uuid>,
    #[serde(default = "default_true")]
    pub include_sql: bool,
    #[serde(default = "default_true")]
    pub include_pipeline_plan: bool,
    pub preferred_provider_id: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CopilotResponse {
    pub answer: String,
    pub suggested_sql: Option<String>,
    pub pipeline_suggestions: Vec<String>,
    pub ontology_hints: Vec<String>,
    pub cited_knowledge: Vec<KnowledgeSearchResult>,
    pub provider_name: String,
    pub cache: SemanticCacheMetadata,
    pub usage: LlmUsageSummary,
    pub created_at: DateTime<Utc>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvaluateGuardrailsRequest {
    pub content: String,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvaluateGuardrailsResponse {
    pub verdict: GuardrailVerdict,
    pub risk_score: f32,
    pub recommendations: Vec<String>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderBenchmarkRequest {
    pub prompt: String,
    pub system_prompt: Option<String>,
    #[serde(default)]
    pub provider_ids: Vec<Uuid>,
    #[serde(default)]
    pub attachments: Vec<ChatAttachment>,
    #[serde(default)]
    pub rubric_keywords: Vec<String>,
    #[serde(default = "default_benchmark_use_case")]
    pub use_case: String,
    #[serde(default = "default_max_tokens")]
    pub max_tokens: i32,
    #[serde(default)]
    pub require_private_network: bool,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderBenchmarkScore {
    pub quality: f32,
    pub latency: f32,
    pub cost: f32,
    pub safety: f32,
    pub overall: f32,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderBenchmarkResult {
    pub provider_id: Uuid,
    pub provider_name: String,
    pub network_scope: String,
    pub reply_preview: String,
    pub prompt_tokens: i32,
    pub completion_tokens: i32,
    pub total_tokens: i32,
    pub estimated_cost_usd: f32,
    pub latency_ms: i32,
    pub cache_hit: bool,
    pub guardrail: GuardrailVerdict,
    pub score: ProviderBenchmarkScore,
    pub error: Option<String>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderBenchmarkResponse {
    pub benchmark_group_id: Uuid,
    pub use_case: String,
    pub prompt_excerpt: String,
    pub required_modalities: Vec<String>,
    pub requested_private_network: bool,
    pub recommended_provider_id: Option<Uuid>,
    pub results: Vec<ProviderBenchmarkResult>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub(crate) struct ConversationRow {
    pub id: Uuid,
    pub title: String,
    pub messages: Json<Vec<ChatMessage>>,
    pub provider_id: Option<Uuid>,
    pub last_cache_hit: bool,
    pub last_guardrail_blocked: bool,
    pub created_at: DateTime<Utc>,
    pub last_activity_at: DateTime<Utc>,
}

impl From<ConversationRow> for Conversation {
    fn from(value: ConversationRow) -> Self {
        Self {
            id: value.id,
            title: value.title,
            messages: value.messages.0,
            provider_id: value.provider_id,
            last_cache_hit: value.last_cache_hit,
            last_guardrail_blocked: value.last_guardrail_blocked,
            created_at: value.created_at,
            last_activity_at: value.last_activity_at,
        }
    }
}

fn default_fallback_enabled() -> bool {
    true
}

fn default_attachment_kind() -> String {
    "text".to_string()
}

fn default_temperature() -> f32 {
    0.2
}

fn default_max_tokens() -> i32 {
    1024
}

fn default_true() -> bool {
    true
}

#[allow(dead_code)]
fn default_benchmark_use_case() -> String {
    "chat".to_string()
}

File: openfoundry-go/libs/ai-kernel-go/models/conversation.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// GuardrailFlag is one annotation produced by the guardrail evaluator.
type GuardrailFlag struct {
	Kind     string `json:"kind"`
	Severity string `json:"severity"`
	Excerpt  string `json:"excerpt"`
}

// GuardrailVerdict bundles the evaluator's status + redacted text.
type GuardrailVerdict struct {
	Status       string          `json:"status"`
	RedactedText string          `json:"redacted_text"`
	Blocked      bool            `json:"blocked"`
	Flags        []GuardrailFlag `json:"flags"`
}

// DefaultGuardrailVerdict mirrors Rust impl Default — status=passed,
// blocked=false, empty redacted text + flags.
func DefaultGuardrailVerdict() GuardrailVerdict {
	return GuardrailVerdict{
		Status: "passed",
		Flags:  []GuardrailFlag{},
	}
}

// ChatAttachment is a multimodal attachment on a chat message.
// Default kind="text".
type ChatAttachment struct {
	Kind       string  `json:"kind"`
	Name       *string `json:"name,omitempty"`
	MimeType   *string `json:"mime_type,omitempty"`
	URL        *string `json:"url,omitempty"`
	Base64Data *string `json:"base64_data,omitempty"`
	Text       *string `json:"text,omitempty"`
}

// ChatMessage is one turn in a conversation.
type ChatMessage struct {
	Role             string                  `json:"role"`
	Content          string                  `json:"content"`
	ProviderID       *uuid.UUID              `json:"provider_id"`
	ToolName         *string                 `json:"tool_name"`
	Citations        []KnowledgeSearchResult `json:"citations,omitempty"`
	Attachments      []ChatAttachment        `json:"attachments,omitempty"`
	GuardrailVerdict *GuardrailVerdict       `json:"guardrail_verdict"`
	CreatedAt        time.Time               `json:"created_at"`
}

// LlmUsageSummary is the per-call cost + latency footer.
type LlmUsageSummary struct {
	PromptTokens     int32   `json:"prompt_tokens"`
	CompletionTokens int32   `json:"completion_tokens"`
	TotalTokens      int32   `json:"total_tokens"`
	EstimatedCostUSD float32 `json:"estimated_cost_usd"`
	LatencyMs        int32   `json:"latency_ms"`
	NetworkScope     string  `json:"network_scope"`
	CacheHit         bool    `json:"cache_hit"`
}

// ChatRoutingMetadata explains which providers got picked + why.
type ChatRoutingMetadata struct {
	RequestedPrivateNetwork bool        `json:"requested_private_network"`
	UsedPrivateNetwork      bool        `json:"used_private_network"`
	PrivacyReason           *string     `json:"privacy_reason"`
	CandidateProviderIDs    []uuid.UUID `json:"candidate_provider_ids"`
	RequiredModalities      []string    `json:"required_modalities"`
}

// SemanticCacheMetadata is the cache footer carried on every reply.
type SemanticCacheMetadata struct {
	CacheKey        string  `json:"cache_key"`
	Hit             bool    `json:"hit"`
	SimilarityScore float32 `json:"similarity_score"`
}

// Conversation is the persisted chat row.
type Conversation struct {
	ID                   uuid.UUID     `json:"id"`
	Title                string        `json:"title"`
	Messages             []ChatMessage `json:"messages"`
	ProviderID           *uuid.UUID    `json:"provider_id"`
	LastCacheHit         bool          `json:"last_cache_hit"`
	LastGuardrailBlocked bool          `json:"last_guardrail_blocked"`
	CreatedAt            time.Time     `json:"created_at"`
	LastActivityAt       time.Time     `json:"last_activity_at"`
}

// ConversationSummary is the trimmed list-view row.
type ConversationSummary struct {
	ID                 uuid.UUID  `json:"id"`
	Title              string     `json:"title"`
	LastMessagePreview string     `json:"last_message_preview"`
	ProviderID         *uuid.UUID `json:"provider_id"`
	MessageCount       int32      `json:"message_count"`
	LastCacheHit       bool       `json:"last_cache_hit"`
	LastActivityAt     time.Time  `json:"last_activity_at"`
}

type ListConversationsResponse struct {
	Data []ConversationSummary `json:"data"`
}

// ChatCompletionRequest defaults: fallback_enabled=true,
// temperature=0.2, max_tokens=1024.
type ChatCompletionRequest struct {
	ConversationID        *uuid.UUID       `json:"conversation_id,omitempty"`
	SystemPrompt          *string          `json:"system_prompt,omitempty"`
	UserMessage           string           `json:"user_message"`
	PurposeJustification  *string          `json:"purpose_justification,omitempty"`
	PromptTemplateID      *uuid.UUID       `json:"prompt_template_id,omitempty"`
	PromptVariables       json.RawMessage  `json:"prompt_variables,omitempty"`
	KnowledgeBaseID       *uuid.UUID       `json:"knowledge_base_id,omitempty"`
	PreferredProviderID   *uuid.UUID       `json:"preferred_provider_id,omitempty"`
	Attachments           []ChatAttachment `json:"attachments,omitempty"`
	FallbackEnabled       bool             `json:"fallback_enabled"`
	RequirePrivateNetwork bool             `json:"require_private_network"`
	Temperature           float32          `json:"temperature"`
	MaxTokens             int32            `json:"max_tokens"`
}

func (r *ChatCompletionRequest) UnmarshalJSON(data []byte) error {
	type alias ChatCompletionRequest
	aux := struct {
		FallbackEnabled *bool    `json:"fallback_enabled"`
		Temperature     *float32 `json:"temperature"`
		MaxTokens       *int32   `json:"max_tokens"`
		*alias
	}{
		alias: (*alias)(r),
	}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	if aux.FallbackEnabled == nil {
		r.FallbackEnabled = DefaultFallbackEnabled
	} else {
		r.FallbackEnabled = *aux.FallbackEnabled
	}
	if aux.Temperature == nil {
		r.Temperature = DefaultTemperature
	} else {
		r.Temperature = *aux.Temperature
	}
	if aux.MaxTokens == nil {
		r.MaxTokens = DefaultMaxTokens
	} else {
		r.MaxTokens = *aux.MaxTokens
	}
	return nil
}

type ChatCompletionResponse struct {
	ConversationID   uuid.UUID               `json:"conversation_id"`
	ProviderID       uuid.UUID               `json:"provider_id"`
	ProviderName     string                  `json:"provider_name"`
	Reply            string                  `json:"reply"`
	Citations        []KnowledgeSearchResult `json:"citations"`
	Guardrail        GuardrailVerdict        `json:"guardrail"`
	Cache            SemanticCacheMetadata   `json:"cache"`
	PromptUsed       string                  `json:"prompt_used"`
	CompletionTokens int32                   `json:"completion_tokens"`
	Usage            LlmUsageSummary         `json:"usage"`
	Routing          ChatRoutingMetadata     `json:"routing"`
	CreatedAt        time.Time               `json:"created_at"`
}

// CopilotRequest defaults: include_sql=true, include_pipeline_plan=true.
type CopilotRequest struct {
	Question             string      `json:"question"`
	PurposeJustification *string     `json:"purpose_justification,omitempty"`
	DatasetIDs           []uuid.UUID `json:"dataset_ids,omitempty"`
	OntologyTypeIDs      []uuid.UUID `json:"ontology_type_ids,omitempty"`
	KnowledgeBaseIDs     []uuid.UUID `json:"knowledge_base_ids,omitempty"`
	IncludeSQL           bool        `json:"include_sql"`
	IncludePipelinePlan  bool        `json:"include_pipeline_plan"`
	PreferredProviderID  *uuid.UUID  `json:"preferred_provider_id,omitempty"`
}

type CopilotResponse struct {
	Answer              string                  `json:"answer"`
	SuggestedSQL        *string                 `json:"suggested_sql"`
	PipelineSuggestions []string                `json:"pipeline_suggestions"`
	OntologyHints       []string                `json:"ontology_hints"`
	CitedKnowledge      []KnowledgeSearchResult `json:"cited_knowledge"`
	ProviderName        string                  `json:"provider_name"`
	Cache               SemanticCacheMetadata   `json:"cache"`
	Usage               LlmUsageSummary         `json:"usage"`
	CreatedAt           time.Time               `json:"created_at"`
}

type EvaluateGuardrailsRequest struct {
	Content string `json:"content"`
}

type EvaluateGuardrailsResponse struct {
	Verdict         GuardrailVerdict `json:"verdict"`
	RiskScore       float32          `json:"risk_score"`
	Recommendations []string         `json:"recommendations"`
}

// ProviderBenchmarkRequest defaults: use_case="chat", max_tokens=1024.
type ProviderBenchmarkRequest struct {
	Prompt                string           `json:"prompt"`
	SystemPrompt          *string          `json:"system_prompt,omitempty"`
	ProviderIDs           []uuid.UUID      `json:"provider_ids,omitempty"`
	Attachments           []ChatAttachment `json:"attachments,omitempty"`
	RubricKeywords        []string         `json:"rubric_keywords,omitempty"`
	UseCase               string           `json:"use_case"`
	MaxTokens             int32            `json:"max_tokens"`
	RequirePrivateNetwork bool             `json:"require_private_network,omitempty"`
}

type ProviderBenchmarkScore struct {
	Quality float32 `json:"quality"`
	Latency float32 `json:"latency"`
	Cost    float32 `json:"cost"`
	Safety  float32 `json:"safety"`
	Overall float32 `json:"overall"`
}

type ProviderBenchmarkResult struct {
	ProviderID       uuid.UUID              `json:"provider_id"`
	ProviderName     string                 `json:"provider_name"`
	NetworkScope     string                 `json:"network_scope"`
	ReplyPreview     string                 `json:"reply_preview"`
	PromptTokens     int32                  `json:"prompt_tokens"`
	CompletionTokens int32                  `json:"completion_tokens"`
	TotalTokens      int32                  `json:"total_tokens"`
	EstimatedCostUSD float32                `json:"estimated_cost_usd"`
	LatencyMs        int32                  `json:"latency_ms"`
	CacheHit         bool                   `json:"cache_hit"`
	Guardrail        GuardrailVerdict       `json:"guardrail"`
	Score            ProviderBenchmarkScore `json:"score"`
	Error            *string                `json:"error"`
}

type ProviderBenchmarkResponse struct {
	BenchmarkGroupID        uuid.UUID                 `json:"benchmark_group_id"`
	UseCase                 string                    `json:"use_case"`
	PromptExcerpt           string                    `json:"prompt_excerpt"`
	RequiredModalities      []string                  `json:"required_modalities"`
	RequestedPrivateNetwork bool                      `json:"requested_private_network"`
	RecommendedProviderID   *uuid.UUID                `json:"recommended_provider_id"`
	Results                 []ProviderBenchmarkResult `json:"results"`
	CreatedAt               time.Time                 `json:"created_at"`
}

// Conversation defaults — exposed for the HTTP-handler slice.
const (
	DefaultAttachmentKind   = "text"
	DefaultBenchmarkUseCase = "chat"
	DefaultFallbackEnabled  = true
	DefaultIncludeSQL       = true
	DefaultIncludePipeline  = true
	DefaultTemperature      = float32(0.2)
	DefaultMaxTokens        = int32(1024)
)
